docs/root/configuration/http_filters/wasm_filter.rst

```diff
@@ -0,0 +1,1184 @@
+
+Returns :ref:`WasmData <config_http_filters_wasm_WasmData>` pointer which holds the trailer value.
+
+getResponseTrailerPairs
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+   WasmDataPtr getResponseTrailerPairs()
+
+Gets all trailer pairs. This method is effective only when called in :ref:`onResponseTrailer <config_http_filters_wasm_context_object_api_onresponsetrailers>` and
+:ref:`onLog <config_http_filters_wasm_context_object_api_onlog>`.
+
+Returns :ref:`WasmData <config_http_filters_wasm_WasmData>` pointer which holds the trailer pairs.
+
+.. _config_http_filters_wasm_body_api:
+
+Body API
+--------
+
+getRequestBodyBufferBytes
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+   WasmDataPtr getRequestBodyBufferBytes(size_t start, size_t length)
+
+Returns buffered request body. This copies segment of request body. *start* is an integer and supplies the body buffer start index to copy. 
+*length* is an integer and supplies the buffer length to copy. This method is effective when calling from :ref:`onRequestBody <config_http_filters_wasm_context_object_api_onrequestbody>`.
+
+Returns :ref:`WasmData <config_http_filters_wasm_WasmData>` pointer which holds the request body data.
+
+getResponseBodyBufferBytes
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+   WasmDataPtr getResponseBodyBufferBytes(size_t start, size_t length)
+
+Returns buffered response body. This copies segment of response body. *start* is an integer and supplies the body buffer start index to copy.
+*length* is an integer and supplies the buffer length to copy. This method is effective when calling from :ref:`onResponseBody <config_http_filters_wasm_context_object_api_onresponsebody>`.
+
+Returns :ref:`WasmData <config_http_filters_wasm_WasmData>` pointer which holds the response body data.
+
+Metadata API
+------------
+
+TODO: Add metadata related API
+
+.. _config_http_filters_wasm_streaminfo_api:
+
+StreamInfo API
+--------------
+
+getProtocol
+^^^^^^^^^^^
+
+.. code-block:: cpp
+
+   WasmDataPtr getProtocol(StreamType type)
+
+Returns the string representation of HTTP protocol used by the current request.
+The possible values are: HTTP/1.0, HTTP/1.1, and HTTP/2.
+*type* is the stream type with two possible values: StreamType::Request and StreamType::Response.
+The string protocol is returned as :ref:`WasmData <config_http_filters_wasm_WasmData>`.
+
+Timer API
+---------
+
+Timer API is used to set a timer and get current timestamp.
+
+.. _config_http_filters_wasm_setTickPeriodMilliseconds:
+
+setTickPeriodMilliseconds
+^^^^^^^^^^^^^^^^^^^^^^^^^
+.. code-block:: cpp
+
+    void setTickPeriodMilliseconds(uint32_t millisecond)
+
+Set a timer. *millisecond* is tick interval in millisecond. :ref:`onTick <config_http_filters_wasm_context_object_api_ontick>` will be invoked when timer fires.
+
+.. note::
+    Only one timer could be set for each WASM module, so it needs to be multiplexed by events with different tick intervals.
```

Comment:
Post https://github.com/envoyproxy/envoy-wasm/pull/44 one timer per root_id which is vectored to the appropriate RootContext.   Any context can call setTickPeriodMilliseconds and the onTick will come on the corresponding RootContext.